### DISCUSSION TITLE
Serious Usability Concerns in New GitHub Classroom

### DISCUSSION THREAD

**Discussion by ctnelson1997** [community/community#125414] [GitHub Education]
Hi all,

Let me start off by saying that I appreciate the continued development on GitHub Classroom. It has been a great way to give students a practical use of Git, while being to easy to manage student submissions. I am very thankful for this tool.

However, I have some serious concerns about the upcoming changes that could compromise our use of GitHub Classroom altogether. I have opted into the beta for a summer session of my course, and I am already experiencing some heartaches that I want to share.

For one, in the old mechanism, starter code repository visibility was independent of student repository visibility. This was useful as we could post our HWs publicly for anyone on the internet to gain and learn from, while still maintaining private student submissions. Now, with repositories inheriting the visibility of the parent, we cannot. Nonetheless, I can just create a private version, e.g. hw1, and public version, e.g. hw1-public. This is only a minor heartache -- fine.

However, there are 2 serious, borderline critical, usability flaws in the new mechanism that I would like to address.

1. Commits are not squashed. Please be aware that there are educators from across the technology spectrum using GitHub classroom, from K-12 to higher ed, from humanities to computer science. Not squashing commits by default is like handing a pair of scissors by the open blade. Someone who has used scissors before will know better, but those who have not will not. Of course, the documentation warns of this, however I cannot imagine a use case where we would _not_ want commits to be squashed. Just like a pair of scissors would  never be passed with an open blade, commits should be squashed by default with the option to toggle this _off_ rather than _on_

2. Forks are dangerous. Maybe even more dangerous than an open pair of scissors? By using GitHub's forking system as-is, there are many "click-me" buttons that we typically do **not** want to encourage. Specifically, pull requests. 

![image](https://github.com/community/community/assets/37457808/889dd1fe-fc59-4d1f-91fc-848a43e60be4)

For example, the interface encourages a confused user to open a pull request. However, by opening a pull request, you share your code with **everyone**. That assignment, which was supposed to be individual, now has a solution shared with the entire class. It takes a lot of effort to scrub this out, which is challenging even for the more advanced of Git users.

![image](https://github.com/community/community/assets/37457808/59b28681-4016-4e84-8cb0-860f05afa688)

**The larger the class, the more likely this is to happen, and the worse the consequences.**

I would really like to use GitHub Classroom going forward! The benefits of these changes are great: a more git-like experience, being able to make changes to starter code mid-assignment, however, the flaws and potential pitfalls, as they are, are much greater. I think additional work needs to be done to educatify the forking mechanism before rolling out these changes. Thank you for your time!

**Comment by ssardina** [comment]
Hi @ctnelson1997 

I totally share your views (including how great GH classroom is).

Your concerns were partially already rose in #112506 back in March. @RyanHecht replied to that, but I haven't see any improvement/news on that front. 

In that discussion, I also added another major issue I had when I tried the beta fork feature:  the **teaching staff would not be able  to use the starter code repo's issue tracker** anymore. 

It is super useful to use the issue tracker of the starter code to address issues with the assessments, including bugs that students' find, or marking criteria, etc. With the fork approach, students would be able to see all those discussions, rendering them not-usable. :sob: 

I also concur that is super useful to start the student's repo with a **fresh commit history,** rather than carrying out all the commits done by teaching staff (not great, may contain commits messages that teachers may not want to show and it's messier). And the PR issue is also super critical.

I wonder if GH can update us on what the plan is. I am starting a teaching semester soon and as far as I can see, the fork approach would make the whole infrastructure significantly  worse in my case. I hope it can stay as template until we really find a full solution to all these issues.

BTW, I should note that while forks allow students to get updates than in the starter code, nothing forces them to do so. That's why I myself push all changes always, using a script.

**Comment by ctnelson1997** [comment]
https://github.com/orgs/community/discussions/127455/ 

Marking as resolved; glad to hear this feedback was taken!
